File: app/src/main/java/uk/co/ourfriendirony/medianotifier/activities/async/ListChildren.java
package uk.co.ourfriendirony.medianotifier.activities.async;

import android.content.Context;
import android.os.AsyncTask;
import android.view.View;
import android.widget.ListView;
import android.widget.ProgressBar;

import java.lang.ref.WeakReference;
import java.util.List;

import uk.co.ourfriendirony.medianotifier.R;
import uk.co.ourfriendirony.medianotifier.db.Database;
import uk.co.ourfriendirony.medianotifier.activities.viewadapter.ListAdapterSummary;
import uk.co.ourfriendirony.medianotifier.mediaitem.MediaItem;

public class ListChildren extends AsyncTask<String, Void, List<MediaItem>> {
    private final WeakReference<Context> context;
    private final WeakReference<ProgressBar> progressBar;
    private final WeakReference<ListView> listView;
    private final Database db;

    public ListChildren(Context context, ProgressBar progressBar, ListView listView, Database db) {
        this.context = new WeakReference<>(context);
        this.progressBar = new WeakReference<>(progressBar);
        this.listView = new WeakReference<>(listView);
        this.db = db;
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        progressBar.get().setVisibility(View.VISIBLE);
    }

    @Override
    protected List<MediaItem> doInBackground(String... params) {
        String id = params[0];
        return db.readChildItems(id);
    }

    @Override
    protected void onPostExecute(final List<MediaItem> children) {
        progressBar.get().setVisibility(View.GONE);
        ListAdapterSummary adapter = new ListAdapterSummary(context.get(), R.layout.list_item_generic_toggle, children, db);
        listView.get().setAdapter(adapter);
        listView.get().setSelection(children.size());
    }
}

File: app/src/main/kotlin/uk/co/ourfriendirony/medianotifier/activities/async/ListChildren.kt
package uk.co.ourfriendirony.medianotifier.activities.async

import android.content.Context
import android.os.Handler
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.ProgressBar
import uk.co.ourfriendirony.medianotifier.R
import uk.co.ourfriendirony.medianotifier.activities.viewadapter.ListAdapterSummary
import uk.co.ourfriendirony.medianotifier.db.Database
import java.lang.ref.WeakReference

class ListChildren(context: Context, progressBar: ProgressBar?, listView: ListView?, db: Database?, ui: Handler, id: String) : Runnable {
    private val context: WeakReference<Context>
    private val progressBar: WeakReference<ProgressBar?>
    private val listView: WeakReference<ListView?>
    private val db: Database?
    private val ui: Handler
    private val id: String

    init {
        this.context = WeakReference(context)
        this.progressBar = WeakReference(progressBar)
        this.listView = WeakReference(listView)
        this.db = db
        this.ui = ui
        this.id = id
    }

    override fun run() {
        ui.post {
            progressBar.get()!!.isIndeterminate = true
        }

        val children = db!!.readChildItems(id)

        ui.post {
            progressBar.get()!!.isIndeterminate = false
            val adapter: ArrayAdapter<*> = ListAdapterSummary(context.get(), R.layout.list_item_generic_toggle, children, db)
            listView.get()!!.adapter = adapter
            listView.get()!!.setSelection(children.size)
        }
    }
}